policy/options.go

```diff
@@ -0,0 +1,702 @@
+package policy
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/net/idna"
+)
+
+type NamePolicyOption func(e *NamePolicyEngine) error
+
+// TODO: wrap (more) errors; and prove a set of known (exported) errors
+
+func WithSubjectCommonNameVerification() NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		e.verifySubjectCommonName = true
+		return nil
+	}
+}
+
+func WithAllowLiteralWildcardNames() NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		e.allowLiteralWildcardNames = true
+		return nil
+	}
+}
+
+func WithPermittedDNSDomains(domains []string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedDomains := make([]string, len(domains))
+		for i, domain := range domains {
+			normalizedDomain, err := normalizeAndValidateDNSDomainConstraint(domain)
+			if err != nil {
+				return errors.Errorf("cannot parse permitted domain constraint %q", domain)
+			}
+			normalizedDomains[i] = normalizedDomain
+		}
+		e.permittedDNSDomains = normalizedDomains
+		return nil
+	}
+}
+
+func AddPermittedDNSDomains(domains []string) NamePolicyOption {
```

Comment:
Instead of doing add and with, is there a reason to not always append with just With*?

Reply:
`google.golang.org/grpc` options that are slices always append, so for example, this add 4 interceptors:
```golang
grpc.DialContext(ctx, target, 
  grpc.WithChainUnaryInterceptor(interceptor1, interceptor2),
  grpc.WithChainUnaryInterceptor(interceptor3),
  grpc.WithChainUnaryInterceptor(interceptor4),
)
```